In [ ]:
모듈 불러오기
1. 사이킷런 데이터셋 - load_iris
2. 사이킷런 트리 - 디시전 트리
3. 사이킷런 모델셀랙션 - 학습테스트분리 , KFold, StratifiedKFold, cross_val_score, cross_validate
4. 사이킷런 매트릭스 - 정확도점수
5. 판다스
6. 넘파이


In [2]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
import pandas as pd

In [22]:
iris=load_iris()

iris


{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [23]:
iris.keys()


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [31]:
i_descr=iris.DESCR
i_data= iris.data
i_label= iris.target
i_tar_name=iris.target_names
i_fea_name=iris.feature_names

print(f'DESCR : {i_descr}')


DESCR : .. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===

In [44]:
print(f'===dataset의 구성=== \n데이터({len(i_data)}개) :{i_data[:5]}\n레이블({len(i_label)}개) :{i_label}\n레이블 네임 : {i_tar_name}\n피처 네임 : {i_fea_name}')


===dataset의 구성=== 
데이터(150개) :[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
레이블(150개) :[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
레이블 네임 : ['setosa' 'versicolor' 'virginica']
피처 네임 : ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


 ---  
 ---  
 # iris data 실습  
 - Estimator : Classified ( DecisionTree)  
     > 1. 일반적인 분류  
     > 2. 분류 + 교차검증 (KFord, StratifiedKFold, cross_val_store())
     > 3. 분류 + 교차검증 + 하이퍼파라미터튜닝 (GridSerachCV)  
---
---

### 1. 일반적인 분류 (DecisionFree활용)  
- 데이터 준비하기 
- 의사결정트리 클래스를 객체로 생성
- train/ test 데이터 분리
- 의사결정트리에 학습용 데이터,레이블 학습
- test용 데이터 사용하여 예측
- 평가

In [64]:
from sklearn.metrics import accuracy_score

#데이터 준비
iris=load_iris()

#의사결정트리 클래스를 객체(clf)로 생성
clf=DecisionTreeClassifier()

#train/ test 데이터 분리
X_train,X_test,y_train,y_test=train_test_split(
    iris.data, iris.target, test_size=0.3, random_state=121)

#의사결정트리에 학습용 피처,레이블 학습
clf.fit(X_train, y_train)

#test용 데이터 예측
pred=clf.predict(X_test)

#평가: 성능평가지표 
print('성능평가지표 :',accuracy_score(y_test,pred))
print('성능평가지표 : {0:.4f}'.format(accuracy_score(y_test,pred)))


성능평가지표 : 0.9555555555555556
성능평가지표 : 0.9556


---
---
### 2. 분류 + 교차검증  
    1) KFord  
    2) StratifiedKFold 
    3) cross_val_score()
- 

    1) KFold : 피처, 레이블이 인덱스로 반환됨 (train index / test index)

In [79]:
from sklearn.model_selection import KFold
import numpy as np

#iris=load_iris()
features=iris.data
label=iris.target
clf=DecisionTreeClassifier(random_state=156)

#kfold객체와 세트별 성능평가(정확도)를 담을 리스트 생성
kfold=KFold(n_splits=5)
cv_accuracy=[]

#for문 사용해서 폴드별(인덱스)로 학습/검증 실행
for train_index, test_index in kfold.split(features):

    #실행별 test_index 확인
    print(f'test_index : {test_index[[0,-1],]}')
    
    #split으로 반환된 인덱스를 활용해 학습/검증용 테스트 데이터 추출
    X_train,X_test=features[train_index], features[test_index]
    y_train,y_test=label[train_index], label[test_index]
    
    #학습
    clf.fit(X_train, y_train)
    #예측
    pred=clf.predict(X_test)
    
    #반복 마다 정확도 측정
    accuracy=np.round(accuracy_score(y_test,pred),4)
    
    #폴드별 정확도를 리스트에 추가
    cv_accuracy.append(accuracy)
    
print(f'폴드 별 정확도 : {cv_accuracy}\n평균 정확도 : {np.mean(cv_accuracy)}')

test_index : [ 0 29]
test_index : [30 59]
test_index : [60 89]
test_index : [ 90 119]
test_index : [120 149]
폴드 별 정확도 : [1.0, 0.9667, 0.8667, 0.9333, 0.7333]
평균 정확도 : 0.9


    2) StratifiedKFold : test데이터의 레이블 비율(산포도)을 전체 데이터의 비율과 동일하게 구성
            for문 사용시 features, label 모두 사용해야함

In [86]:
from sklearn.model_selection import StratifiedKFold

#iris=load_iris()
# features=iris.data
# label=iris.target
#clf=DecisionTreeClassifier(random_state=156)

#StratifiedKFold객체와 세트별 성능평가(정확도)를 담을 리스트 생성
skf=StratifiedKFold(n_splits=5)
cv_accuracy2=[]

for train_idx, test_idx in skf.split(features,label):
    
    print(f'test_index : {test_idx}')

    #split으로 반환된 인덱스를 활용해 학습/검증용 테스트 데이터 추출
    X_train, X_test= features[train_idx],features[test_idx]
    y_train, y_test= label[train_idx],label[test_idx]
    
    #학습
    clf.fit(X_train, y_train)
    
    #평가
    pred=clf.predict(X_test)
    
    #폴드마다 정확도 측정
    accuracy2= np.round(accuracy_score(y_test,pred),4)
    
    #리스트에 정확도 저장
    cv_accuracy2.append(accuracy2)

    
print(f'폴드 별 정확도 : {cv_accuracy2}\n평균 정확도 : {np.mean(cv_accuracy2)}')

test_index : [  0   1   2   3   4   5   6   7   8   9  50  51  52  53  54  55  56  57
  58  59 100 101 102 103 104 105 106 107 108 109]
test_index : [ 10  11  12  13  14  15  16  17  18  19  60  61  62  63  64  65  66  67
  68  69 110 111 112 113 114 115 116 117 118 119]
test_index : [ 20  21  22  23  24  25  26  27  28  29  70  71  72  73  74  75  76  77
  78  79 120 121 122 123 124 125 126 127 128 129]
test_index : [ 30  31  32  33  34  35  36  37  38  39  80  81  82  83  84  85  86  87
  88  89 130 131 132 133 134 135 136 137 138 139]
test_index : [ 40  41  42  43  44  45  46  47  48  49  90  91  92  93  94  95  96  97
  98  99 140 141 142 143 144 145 146 147 148 149]
폴드 별 정확도 : [0.9667, 0.9667, 0.9, 0.9667, 1.0]
평균 정확도 : 0.9600200000000001


    3) cross_val_score() : 폴드세트구분, for문 반복(인덱스추출), 학습-예측-평가지표 과정을 한번에 진행
        estimator : classifier 입력시 stratifiedKFold 실행
                    regressor 입력시 KFold 실행

In [93]:
from sklearn.model_selection import cross_val_score, cross_validate


#iris=load_iris()
# features=iris.data
# label=iris.target
#clf=DecisionTreeClassifier(random_state=156)

#cross_val_score 객체 생성 (estimator, X, y, scoring, cv)
        # estimator: 분류/회귀 알고리즘 , X:피처 , y:레이블 , scoring:성능평가지표 , cv= 교차검증
score=cross_val_score(clf, features, label, scoring='accuracy',cv=5)

#교차검증 별 정확도
print(score)

#평균 정확도
print(np.round(np.mean(score),4))

[0.96666667 0.96666667 0.9        0.96666667 1.        ]
0.96


### 교차 검증의 정리

- 일반 결정트리 정확도 : 0.9556  
  
  
- 교차검증 실행  
    >- KFold n_split(5) : 0.9  
    >- StratifiedKFold n_split(5) : 0.9600  
    >- cross_val_score cv5 : 0.9600  
    
  
  ---
  ---